Make it so name trackers require approval to send spam notification to observer.

pellaea opened the pull request:
Adds name_trackers.approved column, defaulting to true for all existing name_trackers with non-empty note_template.

When new name_trackers are created with a note_template, it starts as unapproved, and sends a notification to the admins.  The admin notification has a link to approve the name_tracker.

All unapproved name_tracker notifications to the observer are ignored.

A message is displayed for the tracker letting them know that their request is pending approval.

They can check if it has been approved in "Your Interests".

-----

Manual test:

1. go to any random observation
2. go to "About xxx" (name page)
3. click on "Email Tracking" in upper right
4. check the "Send a research request to the observer? " checkbox to enable the spammy half of name tracking, then submit the form

You should see a green flash notice letting you know that your request to send notifications to observers is pending approval, along with a link to "Your Interests" where you can see all of your name_tracker notification requests.  On that page you should see your new name_tracker with the note "(Observer notification pending approval.) "

5. grab the name_tracker id from hovering over the "Change to Ignore" or "Destroy" links (not the "Show Name Tracker" link -- that one has the name id instead of the name_tracker id)
6. type in http://localhost:3000/name/approve_tracker/<id> (inserting the appropriate id from 5)
7. now go back to "Your Interests"

The "pending approval" business shouldn't be there any more.  Note that it does a flash notice "Name stalker approved" when an admin approves of a name_tracker.  If you are logged in as a user who is not an admin it will fail.  But you do not have to actually be in admin mode when you request the /name/approve_tracker action.

pellaea committed:
First draft of adding name_trackers.approved column.

pellaea committed:
Name tracker approval appears to be working now.

pellaea committed:
Expanded UI on the "Your Interests" page.

pellaea committed:
NameTracker#approved is ready to go.

pellaea committed:
oops

pellaea commented:
How important is it to send a notification to the tracker when we approve their request?  I guess if we make it a very short, simple email, ASCII-only, it shouldn't be that much work...

JoeCohen commented:
> How important is it to send a notification to the tracker when we approve their request? I guess if we make it a very short, simple email, ASCII-only, it shouldn't be that much work...

Maybe nice to have, but not critical. A reasonable alternative is:
- manually email the user about approval (or rejection). (If/when we reject, we should explain.)
- include a reminder to email the user when the admin is notified about the new request.

JoeCohen commented:
Seems like I have forgotten everything.
I wanted to try a short manual test, but cannot login. What am I doing wrong?
>`rails db:migrate`
`rails lang:update`
`rails s -b 0.0.0.0`
Tried logging in with various combinations of my user name, email, login and "password" || my real password

JoeCohen commented:
Never mind. I forgot to load the db. (At least I was right about forgetting everything.)

JoeCohen commented:
Should `app/views/interests/index.html.erb` display something like "Pending Approval" if a Notification Request is not (yet) approved?

pellaea commented:
Yes, it should display something about approval pending if the name_tracker
is not approved yet.

On Tue, Dec 27, 2022 at 7:36 AM Joseph D. Cohen ***@***.***>
wrote:

> Should app/views/interests/index.html.erb display something like "Pending
> Approval" if a Notification Request is not (yet) approved?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1252#issuecomment-1365871706>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTNNNIYUOFGMPMSIF6SUDWPLPDZANCNFSM6AAAAAATJ6F2AU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

pellaea commented:
Added instructions for manual test in comment.

Also, note the activity log-style interest-type buttons that are now shown on the "Your Interests" index.  If you don't have any interests, then it probably won't show any.  Try using the switch-users feature to my account (login=jason) and see what my interests index looks like for a nontrivial example.   Play around with those button to make sure they work right and that paginations continues to work in each case.

pellaea committed:
Added queued approval email to notify researchers when their name_tracker request is approved.

pellaea committed:
oops

pellaea committed:
oops

JoeCohen reviewed (approved):
Looks good!
Thanks for the manual test. I ran through it, including the additional instructions.

pellaea commented:
Good idea, thanks, Joe.

On Tue, Dec 27, 2022 at 4:55 PM Joseph D. Cohen ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> Looks good!
> Thanks for the manual test. I ran through it, including the additional
> instructions.
> ------------------------------
>
> In app/controllers/name_controller.rb
> <https://github.com/MushroomObserver/mushroom-observer/pull/1252#discussion_r1057622276>
> :
>
> >
>      user = name_tracker.user
> -    name = Name.find(name_tracker.obj_id)
> +    name = name_tracker.name
>      note = name_tracker.note_template
>      subject = "New Name Tracker with Template"
>      content = "User: ##{user.id} / #{user.login}\n" \
>                "Name: ##{name.id} / #{name.search_name}\n" \
>
> Suggestion: For convenience, include user email (to make it marginally
> easier for admins to send a custom response).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1252#pullrequestreview-1230604697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTNNNSCOPFRTRIGVDGRN3WPNQUFANCNFSM6AAAAAATJ6F2AU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

pellaea committed:
Add email address to admin notification for new name_tracker requests.

pellaea commented on app/controllers/name_controller.rb:
```diff
@@ -1137,17 +1137,43 @@ def destroy_name_tracker_interest_and_flash
     note = name_tracker.note_template
     subject = "New Name Tracker with Template"
     content = "User: ##{user.id} / #{user.login}\n" \
               "Name: ##{name.id} / #{name.search_name}\n" \
```
Done, I'll check it in in a minute.

JoeCohen committed:
Convert WebmasterMailer.build to use kwargs

First step in implementing https://www.pivotaltracker.com/story/show/184120794,
which is 1st step to fix ABC offense in `notify_admins_of_name_tracker`

JoeCohen committed:
Update EmailsController calls to WebmasterMailer.build

JoeCohen committed:
Update HerbariaController call to WebmasterMailer.build

JoeCohen committed:
Update LocationController call to WebmasterMailer.build

JoeCohen committed:
Update NameController call to WebmasterMailer.build

JoeCohen committed:
Update Verifications cntlr call to WebmasterMailer.build

JoeCohen committed:
Update CuratorRequests call to WebmasterMailer.build

JoeCohen committed:
Update CreateAndEditName call to WebmasterMailer.build

JoeCohen committed:
Update Name.rb call to WebmasterMailer.build

JoeCohen committed:
Update ApplicationMailerTest call to WebmasterMailer.build

JoeCohen committed:
Update Comment::Callbacks call to WebmasterMailer.build

JoeCohen commented:
Argh! All that work and there's still an ABC error. So close  <2, 20, 3> 20.32/20.
I will simply disable the cop for the offending method.

JoeCohen committed:
Disable ABC cop for notify_admins_of_name_tracker

Fixes Codeclimate issue

JoeCohen commented:
This is another one I'd like to merge if there are no objections.

nimmolo reviewed (approved):
Approve!

pellaea commented:
Yep, thank you, Joe!

On Thu, Dec 29, 2022 at 5:42 PM andrew nimmo ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> Approve!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MushroomObserver/mushroom-observer/pull/1252#pullrequestreview-1232798044>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTNNPIXSZ2WPGQWOS34FTWPYHUDANCNFSM6AAAAAATJ6F2AU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

JoeCohen committed:
Merge branch 'master' into spam-notifications-require-approval

JoeCohen committed:
Update schema by migrating

Fixes selecting the wrong side in resolving merger conflict.
